java/fury-core/src/main/java/org/apache/fury/serializer/StringSerializer.java

```diff
@@ -625,4 +743,534 @@ public String readUTF8String(MemoryBuffer buffer) {
+    } else {
+      // TODO: off-heap
+    }
+    buffer._unsafeWriterIndex(writerIndex);
+  }
+
+  public void writeCharsUTF16(MemoryBuffer buffer, char[] chars) {
+    int numBytes = MathUtils.doubleExact(chars.length);
+    int writerIndex = buffer.writerIndex();
+    buffer.ensure(writerIndex + 5 + numBytes);
+    final byte[] targetArray = buffer.getHeapMemory();
+    if (targetArray != null) {
+      int arrIndex = buffer._unsafeHeapWriterIndex();
+      buffer.putByte(arrIndex, UTF16);
+      buffer.putInt32(arrIndex + 1, numBytes);
+      arrIndex += 5;
+      writerIndex += 5 + numBytes;
+      if (Platform.IS_LITTLE_ENDIAN) {
+        // FIXME JDK11 utf16 string uses little-endian order.
+        Platform.UNSAFE.copyMemory(
+            chars,
+            Platform.CHAR_ARRAY_OFFSET,
+            targetArray,
+            Platform.BYTE_ARRAY_OFFSET + arrIndex,
+            numBytes);
+      } else {
+        heapWriteCharsUTF16BE(chars, arrIndex, numBytes, targetArray);
+      }
+    } else {
+      // TODO: off-heap
+    }
+    buffer._unsafeWriterIndex(writerIndex);
+  }
+
+  public void writeCharsUTF8(MemoryBuffer buffer, char[] chars) {
+    int estimateMaxBytes = chars.length * 3;
+    int writerIndex = buffer.writerIndex();
+    buffer.ensure(writerIndex + 9 + estimateMaxBytes);
+    byte[] targetArray = buffer.getHeapMemory();
+    if (targetArray != null) {
+      int arrIndex = buffer._unsafeHeapWriterIndex();
+      int targetIndex = fastEncodeUTF8(chars, targetArray, arrIndex + 9);
+      int written = targetIndex - arrIndex - 9;
+      buffer.putByte(arrIndex, UTF8);
+      buffer.putInt32(arrIndex + 1, chars.length << 1);
+      buffer.putInt32(arrIndex + 5, written);
+      buffer._unsafeWriterIndex(targetIndex);
+    } else {
+      // TODO: off-heap
+    }
+  }
+
+  public void writeBytesUTF8(MemoryBuffer buffer, byte[] bytes) {
+    int estimateMaxBytes = bytes.length / 2 * 3;
+    int writerIndex = buffer.writerIndex();
+    buffer.ensure(writerIndex + 9 + estimateMaxBytes);
+    byte[] targetArray = buffer.getHeapMemory();
+    if (targetArray != null) {
+      int arrIndex = buffer._unsafeHeapWriterIndex();
+      int targetIndex = fastEncodeUTF8(bytes, targetArray, arrIndex + 9);
+      int written = targetIndex - arrIndex - 9;
+      buffer.putByte(arrIndex, UTF8);
+      buffer.putInt32(arrIndex + 1, bytes.length);
+      buffer.putInt32(arrIndex + 5, written);
+      buffer._unsafeWriterIndex(targetIndex);
+    }
+  }
+
+  private static boolean fastDecodeUTF8(byte[] src, int offset, int len, byte[] dst) {
+    final int end = offset + len;
+    int dp = 0;
+
+    while (offset < end) {
+      if (offset + 8 <= end
+          && (Platform.getLong(src, Platform.BYTE_ARRAY_OFFSET + offset) & 0x8080808080808080L)
+              == 0) {
+        // ascii only
+        for (int i = 0, pos = Platform.IS_LITTLE_ENDIAN ? dp : dp + 1; i < 8; ++i, pos += 2) {
```

Comment:
Could we try to unroll this loop into 8 lines of code and test whether will it be faster

Reply:
Yes, I checked the assembly code and it is indeed not loop-unrolled.